### PR TITLE
Multisigned transactions

### PR DESCRIPTION
**Pull request by TolikH**
# Checklist:

- [ ] Code follows the style guidelines
- [ ] Documentation updated
- [ ] Corner cases evaluated and handled
- [ ] index.js exports updated
- [ ] Automatic tests added/updated
- [ ] Automatic tests pass
- [ ] Sanity check passed
- [ ] Self-review passed
- [ ] Version updated

### CODE REVIEW

File: contracts/Multi-signature/ApprovalCollector.sol [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        require(_criticalApprovers[msg.sender]);
+        _administrators[approver] = false;
+    }
+
+    function addCriticalApprover(address newApprover) public onlyOwner {
+        _criticalApprovers[newApprover] = true;
+    }
+
+    function deleteCriticalApprover(address approver) public onlyOwner {
+        _administrators[approver] = false;
```
Comment (vanruch) [inline contracts/Multi-signature/ApprovalCollector.sol]: I guess it should be `_criticalApprovers[approver] = false;`. Also I'm not sure why critical approver is not always an administrator

File: contracts/Multi-signature/MultiplexingContract.sol [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import "../Consensus/ValidatorSet.sol";
+
+
+contract MultiplexingContract is Ownable {
+
+    address[] private _controlledContracts;
+    mapping (address => uint) private _indexes;
+
+    function performTransaction(address executor, bytes memory transaction) public {
+        bool res = executor.call(transaction);
```
Comment (krzysztof-jelski) [inline contracts/Multi-signature/MultiplexingContract.sol]: For now please just remove this file and its tests and include this line in the ApprovalCollector contract. Let's focus on getting the multisig contract right, and then we'll go to implementing multiplexing.

File: contracts/Multi-signature/ApprovalCollector.sol [hunk 3]
```diff
@@ -0,0 +1,215 @@
+pragma solidity ^0.4.24;
+
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "./MultiplexingContract.sol";
+
+
+/**
+ * @title DelegateERC20 smart contract
+ * @dev Controls execution of ERC20 token transactions
```
Comment (krzysztof-jelski) [inline contracts/Multi-signature/ApprovalCollector.sol]: Remove misleading comment.

File: contracts/Multi-signature/ApprovalCollector.sol [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+ */
+contract ApprovalCollector is Ownable {
+    using SafeMath for uint256;
+
+    MultiplexingContract private multiplexingContract;             //address of controlled contract
+    uint256 constant CALL_LIFETIME = 300000;    //transaction lifetime in seconds
```
Comment (krzysztof-jelski) [inline contracts/Multi-signature/ApprovalCollector.sol]:
```suggestion
    uint256 constant TRANSACTION_LIFETIME = 300000; // in seconds
```
For the value of the constant, please confirm with Vlad.

File: contracts/Multi-signature/ApprovalCollector.sol [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    }
+
+    /**
+     * @dev Adding transaction for approve
+     *
+     * @param executor Address of contract that should execute transaction
+     * @param transaction encoded function selector and parameters
+     * @return identifier of new transaction that should be approved
+     */
+    function executeTransaction( address executor, bytes memory transaction) public {
```
Comment (krzysztof-jelski) [inline contracts/Multi-signature/ApprovalCollector.sol]: `executeTransaction` -> `addTransaction` or `requestTransactionExecution`
param `transaction` -> `callData`

File: contracts/Multi-signature/ApprovalCollector.sol [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    mapping (bytes32 => Transaction) private _transactions;
+
+    mapping (address => bool) private _administrators;
+    mapping (address => bool) private _criticalApprovers;
+
+    mapping (address => ContractClass) private _contractClass;
+    mapping (uint => ContractClassParams) private _contractClassParams;
+
+    ////////////////////////////////////////////////////////////////////////////////////
+    event TransactionCreated(bytes32 transactionId);                            //new transaction created
```
Comment (krzysztof-jelski) [inline contracts/Multi-signature/ApprovalCollector.sol]: Event names are good here. Comments are redundant.

File: contracts/Multi-signature/ApprovalCollector.sol (hunk 5 above)
Comment (krzysztof-jelski) [inline contracts/Multi-signature/ApprovalCollector.sol]: All calls will be performed on the multiplexing contract, so we don't need `executor` parameter.

File: contracts/Multi-signature/ApprovalCollector.sol [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+    /**
+     * @dev Adding transaction for approve
+     *
+     * @param executor Address of contract that should execute transaction
+     * @param transaction encoded function selector and parameters
+     * @return identifier of new transaction that should be approved
+     */
+    function executeTransaction( address executor, bytes memory transaction) public {
+        bytes32 transactionId = keccak256(abi.encodePacked(executor, _transactionNonce, transaction));
```
Comment (krzysztof-jelski) [inline contracts/Multi-signature/ApprovalCollector.sol]: Extract method.

File: contracts/Multi-signature/ApprovalCollector.sol [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+        uint256 timestamp;
+        uint arrayIndex;    //index in array of pending transactions
+
+        ContractClassParams classParams;
+
+        bytes transaction;  //function selector + encoded arguments
+        address executor;   //contract, that should perform call
+    }
+
+    bytes32[] _pendingTransactions;
```
Comment (krzysztof-jelski) [inline contracts/Multi-signature/ApprovalCollector.sol]: In all other contracts we don't prefix members with `_`. It would be good to keep this style consistent.

File: contracts/Multi-signature/ApprovalCollector.sol [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+    mapping (bytes32 => Transaction) private _transactions;
+
+    mapping (address => bool) private _administrators;
+    mapping (address => bool) private _criticalApprovers;
```
Comment (krzysztof-jelski) [inline contracts/Multi-signature/ApprovalCollector.sol]: I think the code would be more readable if the names had more symmetry. For example, `administrators - criticalAdministrators` or `approvers - criticalApprovers`.

File: contracts/Multi-signature/ApprovalCollector.sol [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+    /**
+     * @dev Function for approving transaction. If this is a last needed approval that executes transaction
+     * Reverts if sender hasn't permission to approve transaction, transaction isn't exists or already performed,
+     * transaction timeout or transaction already approved
+     *
+     * @param transactionId Identifier of transaction to approve
+     */
+    function approveTransaction(bytes32 transactionId) public onlyAdmin {
+        require(!_hasApproved(msg.sender, transactionId), "Transaction already approved");
+        require(_transactions[transactionId].timestamp != 0, "Transaction isn\'t exist or already performed");
```
Comment (krzysztof-jelski) [inline contracts/Multi-signature/ApprovalCollector.sol]:
```suggestion
        require(_transactions[transactionId].timestamp != 0, "Transaction doesn\'t exist or was already performed");
```

File: contracts/Multi-signature/ApprovalCollector.sol [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+    function addCriticalApprover(address newApprover) public onlyOwner {
+        _criticalApprovers[newApprover] = true;
+    }
+
+    function deleteCriticalApprover(address approver) public onlyOwner {
+        _administrators[approver] = false;
+    }
+
+    function setContractClass(address _contract, ContractClass class) public onlyAdmin {
+        _contractClass[_contract] = class;
```
Comment (spherefoundry) [inline contracts/Multi-signature/ApprovalCollector.sol]: By only limiting this method to onlyAdmin, an "regular" admin can lower the contract class and execute a transaction that would normally require an "critical" admin.

File: contracts/Multi-signature/ApprovalCollector.sol [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+    }
+
+    /**
+     * @dev Function for getting current controlled contract
+     */
+    function getMultiplexingContract() public view returns (MultiplexingContract) {
+        return multiplexingContract;
+    }
+
+    function updateMultiplexingContract(MultiplexingContract newContractAddress) public onlyOwner {
```
Comment (spherefoundry) [inline contracts/Multi-signature/ApprovalCollector.sol]: This should not be possible at all, or at least require a vote. Otherwise, Admins could vote on one transaction to be executed, and then any admin (even the lower ones) could replace the contract being executed.

File: contracts/Multi-signature/ApprovalCollector.sol (hunk 5 above)
Comment (spherefoundry) [reply contracts/Multi-signature/ApprovalCollector.sol]: I would extend on this: the multiplexer contract is supposed to take care of the fact different transactions go to different contracts. But on the other hand we want be able to control the `execution class` of transactions. Instead of using the target contract address, we could use the method signature (first 4 bytes of Call-Data)

File: contracts/Multi-signature/ApprovalCollector.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (TolikH) [reply contracts/Multi-signature/ApprovalCollector.sol]: Fixed

File: contracts/Multi-signature/ApprovalCollector.sol [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+        _administrators[newAdmin] = true;
+    }
+
+    function deleteAdministrator(address approver) public {
+        require(_criticalApprovers[msg.sender]);
+        _administrators[approver] = false;
+    }
+
+    function addCriticalApprover(address newApprover) public onlyOwner {
+        _criticalApprovers[newApprover] = true;
```
[the remark this answers is not in the document]
Comment (TolikH) [reply contracts/Multi-signature/ApprovalCollector.sol]: Done

File: test/contracts/front/approve_collector.js [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+
+    await expect(deleteAdministrator(totalStranger)).to.be.fulfilled;
+    await expect(approveTransaction(transactions[1], totalStranger)).to.be.eventually.rejected;
+  });
+
+  it('Delete administrator without permissions', async () => {
+    await expect(deleteAdministrator(totalStranger, other)).to.be.eventually.rejected;
+  });
+
+  it('Perform transaction', async () => {
```
[the remark this answers is not in the document]
Comment (TolikH) [reply test/contracts/front/approve_collector.js]: Added new event that returns result of transaction execution

File: contracts/Multi-signature/MultiplexingContract.sol (hunk 2 above)
Comment (TolikH) [reply contracts/Multi-signature/MultiplexingContract.sol]: I wasn't able to remove file, because I need real contract interface for creating stub . I'll create stub and update tests right after real multiplexing contract implementation. For now I only removed all logic from contract.

File: contracts/Multi-signature/ApprovalCollector.sol (hunk 3 above)
Comment (TolikH) [reply contracts/Multi-signature/ApprovalCollector.sol]: Done

File: contracts/Multi-signature/ApprovalCollector.sol (hunk 4 above)
Comment (TolikH) [reply contracts/Multi-signature/ApprovalCollector.sol]: Done

File: contracts/Multi-signature/ApprovalCollector.sol (hunk 5 above)
Comment (TolikH) [reply contracts/Multi-signature/ApprovalCollector.sol]: I changed to addTransaction

File: contracts/Multi-signature/ApprovalCollector.sol (hunk 8 above)
Comment (TolikH) [reply contracts/Multi-signature/ApprovalCollector.sol]: Fixed

File: contracts/Multi-signature/ApprovalCollector.sol (hunk 10 above)
Comment (TolikH) [reply contracts/Multi-signature/ApprovalCollector.sol]: Fixed

File: contracts/Multi-signature/ApprovalCollector.sol [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+        );
+    }
+
+    function addAdministrator(address newAdmin) public {
+        require(_criticalApprovers[msg.sender]);
+        _administrators[newAdmin] = true;
+    }
+
+    function deleteAdministrator(address approver) public {
+        require(_criticalApprovers[msg.sender]);
```
[the remark this answers is not in the document]
Comment (TolikH) [reply contracts/Multi-signature/ApprovalCollector.sol]: Added modifier
